classes/core/PKPContainer.inc.php

```diff
@@ -272,6 +273,43 @@ protected static function getDefaultMailer(): string
 
         return $default;
     }
+
+    /**
+     * Setting a proxy on the stream_context_set_default when configuration [proxy] is filled
+     */
+    protected function settingProxyForStreamContext(): void
+    {
+        $proxyUri = null;
+
+        if ($httpProxy = Config::getVar('proxy', 'http_proxy')) {
+            $proxyUri = $httpProxy;
+        }
+
+        if ($httpsProxy = Config::getVar('proxy', 'https_proxy')) {
+            $proxyUri = $httpsProxy;
+        }
+
+        if (!$proxyUri) {
+            return;
+        }
+
+        /**
+         * `Connection close` here its to avoid slowness. More info at https://www.php.net/manual/en/context.http.php#114867
+         * `request_fulluri` its related to avoid proxy errors. More info at https://www.php.net/manual/en/context.http.php#110449
+         */
+        $opts = [
+            'http' => [
```

Comment:
I would expect a `http` proxy to be configured in an `http` section, and an `https` proxy to be configured in a separate `https` section, rather than an `http` proxy taking precedence over `https` and only configuring the `http` section of the options. It does appear there may be an open underlying PHP bug related to this (https://bugs.php.net/bug.php?id=74796) that might affect testing, though.

Reply:
According to those comments [[1]](https://www.php.net/manual/en/function.stream-context-create.php#74795) [[2]](https://www.php.net/manual/en/function.stream-context-create.php#74795) the key should be `http` even if we pass a HTTPS proxy.

This is mentioned on a comment at your link too.

```
 [2017-06-22 11:42 UTC] anrdaemon at freemail dot ru
And I just found out that settings "https" context has no effect.
Both "https://" and "http://" URL's fall into "http" context.
WHY?…
```

Reply:
@asmecher /\

Reply:
What form did you write your `http_proxy` / `https_proxy` configurations in `config.inc.php` to get this working?

Reply:
I've filled just only one of them.

Reply:
OK, but what form was the URL?

Reply:
`http_proxy = "http://127.0.0.1:8888"`

Reply:
Hmm, what do you get for `stream_get_transports()`? Is this tested working with Tinyproxy?

Reply:
I've removed the `http://` schema from `http_proxy` URI config.

```
http_prox="127.0.0.1:8888"
```

Used this code on `index.php` to test the xml loading over proxy:

```
function getPluginsXML() {
    $url = 'https://pkp.sfu.ca/ojs/xml/plugins.xml';

    dd(simplexml_load_file($url));
    return true;
}

register_shutdown_function('getPluginsXML');
```


Those are the logs from `/var/log/tinyproxy/tinyproxy.log` when I hit the OJS url.

```
CONNECT   Feb 02 20:01:47 [1496706]: Connect (file descriptor 7): omp [127.0.0.1]
CONNECT   Feb 02 20:01:47 [1496706]: Request (file descriptor 7): CONNECT pkp.sfu.ca:443 HTTP/1.0
INFO      Feb 02 20:01:47 [1496706]: No upstream proxy for pkp.sfu.ca
INFO      Feb 02 20:01:47 [1496706]: opensock: opening connection to pkp.sfu.ca:443
INFO      Feb 02 20:01:48 [1496706]: opensock: getaddrinfo returned for pkp.sfu.ca:443
CONNECT   Feb 02 20:01:48 [1496706]: Established connection to host "pkp.sfu.ca" using file descriptor 8.
INFO      Feb 02 20:01:48 [1496706]: Not sending client headers to remote machine
INFO      Feb 02 20:01:52 [1496706]: Closed connection between local client (fd:7) and remote client (fd:8)
```

And the printed response on front-end:

![Screenshot from 2022-02-02 20-02-52](https://user-images.githubusercontent.com/1243634/152253817-983a4d68-4437-4521-836f-4241c224faa1.png)

Reply:
***Testing***

Have you tested with the original use case? Copying my instructions from Slack before they disappear:

First step: Configure the DOI plugin.
- Go to Settings > Website > Plugins
  - Enable the DOI plugin
  - Go to the DOI plugin's Settings
    - Tick "Articles"
    - Enter e.g. "10.1234" in DOI prefix
    - Save the form
Second step: Assign a DOI to a submission.
- Go to Submissions > Mwandenga
  - Go to Publication
  - Go to Identifiers
  - Assign a DOI (you may have to unpublish first)
  - Publish the submission.
Third step: Export the submission using CrossRef XML.
- Go to the Tools > Import/Export > CrossRef
  - Fill in a name, email, username, and password into the settings form
  - Hit "Save"
  - You should now see the "Articles" tab (reload the page if not)
  - The "Articles" tab should have the published article from the earlier step listed.
  - Select the checkbox for XML validation.
  - Select the checkbox beside the article and hit "Export" to generate the XML.

You should see Tinyproxy log entries showing requests to CrossRef as it fetches the schema files.

***URL formats***

We have two different scenarios in which `http_proxy` / `https_proxy` will be used, and we need to make sure that both uses expect the same format. The right approach needs to be documented in `config.TEMPLATE.inc.php`. If possible we should avoid breaking existing configurations, so if `http` URLs were our previously recommended configuration, then it would be best to stay with it.

The two use cases are:

1. Guzzle (the form that was supported in the past)
    The format is documented here but not in great detail: https://docs.guzzlephp.org/en/stable/request-options.html#proxy
    As a transport, Guzzle uses a few options behind the scenes, e.g. CURL (if available), or PHP's stream functions (if it's not). So the format needs to work for all of Guzzle's transports and we should stay as standard as possible in our recommendations.

2. PHP's default stream context (this is what you're adding).

I checked Guzzle's source and found that [it transforms](https://github.com/guzzle/guzzle/blob/c1fd316f0a0f3325ed1e7cdbe61030418b868f9f/src/Handler/StreamHandler.php#L445) `http://` URLs into `tcp://`. This is in line with what [stream_context_set_default](https://www.php.net/manual/en/function.stream-context-set-default.php) recommends. So I suggest adopting the same pattern rather than dropping the scheme part of the URL.

Reply:
@asmecher went thru all testing steps, and seems everything worked well. Following it's the `tinyproxy.log` excerpt.

![Screenshot from 2022-02-08 00-45-42](https://user-images.githubusercontent.com/1243634/152915279-46bf1b1b-d943-412f-bd40-e4f9dac656d3.png)

Reply:
OK, looks good. Is it ready for a code review?

Reply:
It's ready to another code review round.